pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,816 @@
+
+	c := &CapacityScheduling{
+		frameworkHandle:   handle,
+		elasticQuotaInfos: NewElasticQuotaInfos(),
+		pdbLister:         getPDBLister(handle.SharedInformerFactory()),
+	}
+
+	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	client, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	schedSharedInformerFactory := schedinformer.NewSharedInformerFactory(client, 0)
+	c.elasticQuotaLister = schedSharedInformerFactory.Scheduling().V1alpha1().ElasticQuotas().Lister()
+	elasticQuotaInformer := schedSharedInformerFactory.Scheduling().V1alpha1().ElasticQuotas().Informer()
+	elasticQuotaInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1alpha1.ElasticQuota:
+					return true
+				case cache.DeletedFinalStateUnknown:
+					if _, ok := t.Obj.(*v1alpha1.ElasticQuota); ok {
+						return true
+					}
+					utilruntime.HandleError(fmt.Errorf("cannot convert to *v1.Pod: %v", obj))
+					return false
+				default:
+					utilruntime.HandleError(fmt.Errorf("unable to handle object in %T", obj))
+					return false
+				}
+			},
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addElasticQuota,
+				UpdateFunc: c.updateElasticQuota,
+				DeleteFunc: c.deleteElasticQuota,
+			},
+		})
+
+	schedSharedInformerFactory.Start(nil)
+	if !cache.WaitForCacheSync(nil, elasticQuotaInformer.HasSynced) {
+		return nil, fmt.Errorf("Timed out waiting for caches to sync %v", Name)
+	}
+
+	podInformer := handle.SharedInformerFactory().Core().V1().Pods().Informer()
+	podInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1.Pod:
+					return assignedPod(t)
+				case cache.DeletedFinalStateUnknown:
+					if pod, ok := t.Obj.(*v1.Pod); ok {
+						return assignedPod(pod)
+					}
+					return false
+				default:
+					return false
+				}
+			},
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addPod,
+				DeleteFunc: c.deletePod,
+			},
+		},
+	)
+	klog.Infof("CapacityScheduling start")
+	return c, nil
+}
+
+// PreFilter performs the following validations.
+// 1. Check if the (Pod.request + Quota.Allocated) is less than Quota.Max.
+// 2. Check if the sum(eq's usage) > sum(eq's min).
+func (c *CapacityScheduling) PreFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod) *framework.Status {
+	snapshotElasticQuota := c.snapshotElasticQuota()
+	preFilterState := computePodResourceRequest(pod)
+
+	state.Write(preFilterStateKey, preFilterState)
+	state.Write(ElasticQuotaSnapshotKey, snapshotElasticQuota)
+
+	elasticQuotaInfos := snapshotElasticQuota.elasticQuotaInfos
+	eq := snapshotElasticQuota.elasticQuotaInfos[pod.Namespace]
+	if eq == nil {
+		return framework.NewStatus(framework.Success, "skipCapacityScheduling")
+	}
```

Comment:
> state.Write(preFilterStateKey, preFilterState)
	state.Write(ElasticQuotaSnapshotKey, snapshotElasticQuota)
---
These values will be useful later.

Reply:
`eq == nil` indicates the pod isn't subject to EQ, so isn't it unnecessary to calculate the pre-states?

Reply:
```
    postFilter:
      enabled:
      - name: CapacityScheduling
      disabled:
      - name: "*"
```
Because we use CapacityScheduling replace the defaultpreemption. So although the pod isn't subject to EQ, it also can preempt other pods which also isn't subject to EQ. So pre-states also need to be used in PostFilter.